#!/usr/bin/env python

PACKAGE = 'background_subtraction'

from dynamic_reconfigure.parameter_generator_catkin import *;

gen = ParameterGenerator()

gen.add("use_camera_info", bool_t, 0, "Indicates that the camera_info topic should be subscribed to to get the default input_frame_id. Otherwise the frame from the image message will be used.", True)
algorithm_type = gen.enum([ gen.const("MOG",  int_t, 0, "BackgroundSubtractorMOG"),
                            gen.const("MOG2", int_t, 1, "BackgroundSubtractorMOG2"),
                            gen.const("GMG",  int_t, 2, "BackgroundSubtractorGMG")], "An enum for Background Subtraction Mehtods")

gen.add("algorithm_type", int_t, 0, "Background Subtraction Algorithm Methods", 1, 0, 2, edit_method=algorithm_type)

gen.add("mog_history", int_t, 0, "Length of the history", 200, 0, 2048)
gen.add("mog_nmixtures", int_t, 0, "Number of Gaussian mixtures", 5, 0, 128)
gen.add("mog_background_ratio", double_t, 0, "backgroundRatio", 0.7, 0.0, 1.0)
gen.add("mog_noise_sigma", double_t, 0, "Noise strength", 30*0.5, 0.0, 512.0)
gen.add("mog_learning_rate", double_t, 0, "Learning rate", 0.0, -1.0, 1.0)

gen.add("mog2_history", int_t, 0, "Length of the history", 500, 0, 2048)
gen.add("mog2_var_threshold", double_t, 0, "Threshold on the squared Mahalanobis distance to decide whether it is well described by the background model. This parameter does not affect the background update. A typical value could be 4 sigma, that is, varThreshold=4*4=16;", 16.0, 0.0, 256.0)
gen.add("mog2_shadow_detection", bool_t, 0, "Parameter defining whether shadow detection should be enabled (true or false)", True)
gen.add("mog2_learning_rate", double_t, 0, "Learning rate", -1.0, -1.0, 1.0)
            
gen.add("gmg_max_features", int_t, 0, "Total number of distinct colors to maintain in histogram", 64, 1, 512)
gen.add("gmg_learning_rate", double_t, 0, "Set between 0.0 and 1.0, determines how quickly features are 'forgotten' from histograms", 0.025, 0.0, 1.0);
gen.add("gmg_num_initialization_frames", int_t, 0, "Number of frames of video to use to initialize histograms", 120, 1, 1024)
gen.add("gmg_quantization_levels", int_t, 0, "Number of discrete levels in each channel to be used in histograms", 16, 1, 128)
gen.add("gmg_background_prior", double_t, 0, "Prior probability that any given pixel is a background pixel. A sensitivity parameter", 0.8, 0.0, 1.0)
gen.add("gmg_decision_threshold", double_t, 0, "Value above which pixel is determined to be FG.", 0.8, 0.0, 1.0)
gen.add("gmg_smoothing_radius", double_t, 0, "Smoothing radius, in pixels, for cleaning up FG image.", 7, 1, 32)
gen.add("gmg_update_background_model", bool_t, 0, "Perform background model update", True)


exit (gen.generate (PACKAGE, "background_subtraction", "BackGroundSubtraction"))
